espnet2/bin/enh_inference_streaming.py

```diff
@@ -0,0 +1,545 @@
+
+            if enh_s2t_task:
+                arg_list = ("enh_encoder", "enh_separator", "enh_decoder")
+            else:
+                arg_list = ("encoder", "separator", "decoder")
+            supported_keys = list(chain(*[[k, k + "_conf"] for k in arg_list]))
+            for k in infer_args.keys():
+                if k not in supported_keys:
+                    raise ValueError(
+                        "Only the following top-level keys are supported: %s"
+                        % ", ".join(supported_keys)
+                    )
+
+            recursive_dict_update(train_args, infer_args, verbose=True)
+            enh_train_args = argparse.Namespace(**train_args)
+            enh_model = build_model_from_args_and_file(
+                task, enh_train_args, model_file, device
+            )
+
+        if enh_s2t_task:
+            enh_model = enh_model.enh_model
+        enh_model.to(dtype=getattr(torch, dtype)).eval()
+
+        self.device = device
+        self.dtype = dtype
+        self.enh_train_args = enh_train_args
+        self.enh_model = enh_model
+        self.num_spk = enh_model.num_spk
+
+        self.frame_size = enh_model.encoder.frame_size
+        self.hop_size = enh_model.encoder.hop_size
+
+        task = "enhancement" if self.num_spk == 1 else "separation"
+
+        # reference channel for processing multi-channel speech
+        if ref_channel is not None:
+            logging.info(
+                "Overwrite enh_model.separator.ref_channel with {}".format(ref_channel)
+            )
+            enh_model.separator.ref_channel = ref_channel
+            self.ref_channel = ref_channel
+        else:
+            self.ref_channel = enh_model.ref_channel
+
+        self.streaming_states = None
+
+    def reset(self):
+        self.streaming_states = None
+
+    @torch.no_grad()
+    def __call__(
+        self, speech_mix: Union[torch.Tensor, np.ndarray], fs: int = 8000
+    ) -> List[torch.Tensor]:
+        """Inference
+
+        Args:
+            speech_mix: Input speech data (Batch, Nsamples [, Channels])
+            fs: sample rate
+        Returns:
+            [separated_audio1, separated_audio2, ...]
+
+        """
+        assert check_argument_types()
+
+        # Input as audio signal
+        if isinstance(speech_mix, np.ndarray):
+            speech_mix = torch.as_tensor(speech_mix)
+
+        assert speech_mix.dim() > 1, speech_mix.size()
+        batch_size = speech_mix.size(0)
+        speech_mix = speech_mix.to(getattr(torch, self.dtype))
+        # lengths: (B,)
+        lengths = speech_mix.new_full(
+            [batch_size], dtype=torch.long, fill_value=speech_mix.size(1)
+        )
+
+        # a. To device
+        speech_mix = to_device(speech_mix, device=self.device)
+        lengths = to_device(lengths, device=self.device)
+
+        # b. Enhancement/Separation Forward
+        frame_feature, f_lens = self.enh_model.encoder(speech_mix, lengths)
```

Comment:
Do we need to consider possible buffering in the encoder and decoder as well if we feed the frames one-by-one? I mean if the encoder/decoder is also a multi-layer module, it might be wrong to simply use the default forward function.

Reply:
Certain cases may make this problem critical. For example, if an STFT encoder with `center=True` is used, it will apply zero-paddings to both size and the output will be multiple frames.

Reply:
In cases where a multi-layer encoder/decoder is used, it might be difficult to define an exact `frame_size` or `hop_size`.

Reply:
Add `forward_streaming` for the encoder and decoder now. 

> In cases where a multi-layer encoder/decoder is used, it might be difficult to define an exact frame_size or hop_size.
It can be determined by the ABSEncoder and ABSEncoder abstract function:
```
  @property
  def frame_size(self) -> int:
      raise NotImplementedError

```